packages/ec-core/src/weighted/weighted_pair.rs

```diff
@@ -0,0 +1,179 @@
+        let weighted = Weighted::new(Best, 5).with_item_and_weight(Worst, 8);
+        assert!(matches!(
+            weighted,
+            Ok(WeightedPair {
+                a: Weighted {
+                    item: Best,
+                    weight: 5
+                },
+                b: Weighted {
+                    item: Worst,
+                    weight: 8
+                },
+                distr: Some(_),
+                weight_sum: 13,
+            })
+        ));
+        assert_eq!(13, weighted.unwrap().weight());
+    }
+
+    #[proptest]
+    fn best_or_worst(#[map(|v: [i32;10]| v.into())] pop: Vec<i32>) {
+        let mut rng = rand::thread_rng();
+        // We'll make a selector that has a 50/50 chance of choosing the highest
+        // or lowest value.
+        let weighted = Weighted::new(Best, 1)
+            .with_item_and_weight(Worst, 1)
+            .unwrap();
+        let selection = weighted.select(&pop, &mut rng).unwrap();
+        let extremes: [&i32; 2] = pop.iter().minmax().into_option().unwrap().into();
+        assert!(extremes.contains(&selection));
+    }
+
+    #[proptest]
+    fn several_selectors(#[map(|v: [i32;10]| v.into())] pop: Vec<i32>) {
+        let mut rng = rand::thread_rng();
+        // We'll make a selector that has a 50/50 chance of choosing the highest
+        // or lowest value.
+        let weighted = Weighted::new(Best, 1)
+            .with_item_and_weight(Worst, 1)
+            .with_item_and_weight(Random, 0)
+            .with_item_and_weight(Tournament::of_size::<3>(), 2)
+            .unwrap();
+        let selection = weighted.select(&pop, &mut rng).unwrap();
+        assert!(pop.contains(selection));
+    }
+
+    #[test]
+    fn zero_weight_sum_error() {
+        let pop = vec![5, 8, 9, 6, 3, 2, 0];
+        let mut rng = rand::thread_rng();
+        let weighted = Weighted::new(Best, 0)
+            .with_item_and_weight(Worst, 0)
+            .with_item_and_weight(Random, 0)
+            .unwrap();
+        // If all the weights are zero, then selection should return an appropriate
+        // error type.
+        assert_eq!(
+            weighted.select(&pop, &mut rng).unwrap_err(),
+            SelectionError::from(ZeroWeight)
+        );
+    }
+
+    #[test]
+    fn weight_sum_overflow_error() {
+        let weighted = Weighted::new(Best, u32::MAX).with_item_and_weight(Worst, 0);
+        assert!(weighted.is_ok());
+        let weighted = weighted
+            .with_item_and_weight(Random, u32::MAX - 1)
+            .unwrap_err();
+        assert_eq!(weighted, WeightSumOverflow(u32::MAX, u32::MAX - 1));
+    }
+
+    #[test]
+    fn weight_sum_overflow_error_chaining() {
+        let weighted = Weighted::new(Best, u32::MAX).with_item_and_weight(Random, u32::MAX - 1);
+        assert!(weighted.is_err());
+        let weighted = weighted.with_item_and_weight(Worst, 0).unwrap_err();
+        assert_eq!(weighted, WeightSumOverflow(u32::MAX, u32::MAX - 1));
+    }
+}
```

Comment:
Do we have a simple way to make the underlying selector error and add tests for that? If not we should maybe add a simple 

```rs
struct SelectorError;

struct ErroringSelector;

impl Selector for ErroringSelector {
    type Error = SelectorError;

    fn select(...) -> ... { SelectorError }
}
```
and add tests for that

Reply:
I moved this to a new issue: #269